### PR TITLE
release-23.2: sql: fix sql activity cache for large cardinality

### PR DESCRIPTION
**Pull request by maryliag**
Backport 2/2 commits from #112350.
Backport 1/1 commits from #112891.
Backport 1/1 commits from #112975.

/cc @cockroachdb/release

---

tldr: fix activity update job failing from conflict, caching to much info when their is high cardinatlity in the stats, and fix cache to store the correct top stats.

1. Fix activity update failure from upserting same row multiple times. This was caused by 2 nodes having the same statement, but one only executed it 2 or 3 time and the other node executed it enough to generate an index recommendation. The different index recommendation caused duplicate rows based on primary key to be inserted. It now uses index recommendation from the merged stats which uses the latest existing
recommendation (we don't update if there is no recommendation).
2. Customers can have transactions that have over 1k unique statement fingerprint ids. This makes it unreasonable to cache all of the statement information for all the top transactions. This commit removes the logic to cache all the statement information for the top transactions which could cause the table to grow to over 100k for a single hour. This requires the call to go the statistics table to get all the necessary statement information for the txn details.
3. The activity cache logic was not correctly selecting the top CPU json column. It was using the computed index column name instead of the json field name. The values shown on the UI were correct because the endpoint still aggregated several columns using the aggregated stats rather than the separate column.
4. The activity cache logic was using the execution count instead of the statistics count to select the top queries. In most cases this still generates the correct top queries, and the UI uses the aggregated stats which showed the correct value.

Fixes: #111780
Fixes: #111224

Release note (bug fix): Fix the SQL activity update job to avoid conflicts on update, reduces the amount of data cached to just what the overview page requires, and fix the correctess of the top queries.

---

sql: builtin for jsonb array to string array conversion
This commit introduces the function
`jsonb_array_to_string_array` that converts a JSONB array
to a string array format.

Epic: none

Release note (sql change): New builtin `jsonb_array_to_string_array`
that converts JSONB array to string array.

---

sql: fix for crash with null elements on jsonb_array_to_string_array
Fixes https://github.com/cockroachdb/cockroach/issues/112829

The builtin jsonb_array_to_string_array was crashing, when it
had null elements. This commit handles this case, removing them
from the final array.

Release note: None

---

This commit makes the builtin jsonb_array_to_string_array return NULL elements, which were previously being removed. It also adds more testing for different types of elements on the array.

Release note: None

---

Release justification: bug fixes and performance improvement

### CODE REVIEW

**Review by yuzefovich — APPROVED** [review]
> Reviewed 4 of 4 files at r1, 8 of 8 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

**Review by mgartner — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, 8 of 8 files at r2, 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @maryliag)

___
*[`pkg/sql/sem/builtins/builtins.go` line 3577 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112865#-NhX9hf17zCQYOusc4Qx:-NhX9hf17zCQYOusc4Qy:b-yd1cyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e1158168939266efb00c9c4992f530c8b4e6510/pkg/sql/sem/builtins/builtins.go#L3577)):*
> ```Go
> 				return strArray, nil
> 			},
> 			Info:              "Convert a JSONB array into a string array, removing 'null' elements.",
> ```

Why whould nulls be removed by this built-in? It seems more natural for the JSON `null`s to become SQL `NULL`s.
___
*[`pkg/sql/logictest/testdata/logic_test/array` line 1234 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112865#-NhXAF4e7d98BOG4Cj3w:-NhXAF4e7d98BOG4Cj3x:bca10t3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e1158168939266efb00c9c4992f530c8b4e6510/pkg/sql/logictest/testdata/logic_test/array#L1234)):*
> ```
> 
> query T
> SELECT jsonb_array_to_string_array('["foo", null]':::JSONB)
> ```

We should have some tests with the array elements as all types of JSON: numbers, arrays, objects.

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @mgartner)

___
*[`pkg/sql/sem/builtins/builtins.go` line 3577 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112865#-NhX9hf17zCQYOusc4Qx:-NhXLP6j70XPUCfJbCRP:b-qh4daa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e1158168939266efb00c9c4992f530c8b4e6510/pkg/sql/sem/builtins/builtins.go#L3577)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Why whould nulls be removed by this built-in? It seems more natural for the JSON `null`s to become SQL `NULL`s.
</blockquote></details>

Made the changes on a new PR https://github.com/cockroachdb/cockroach/pull/112975
once that gets approved/merged I will include on this backport
___
*[`pkg/sql/logictest/testdata/logic_test/array` line 1234 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112865#-NhXAF4e7d98BOG4Cj3w:-NhXLVJ50_EPl9frL9Oj:b-ci52bw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e1158168939266efb00c9c4992f530c8b4e6510/pkg/sql/logictest/testdata/logic_test/array#L1234)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

We should have some tests with the array elements as all types of JSON: numbers, arrays, objects.
</blockquote></details>

Tests added on the PR mentioned above

**Review by dhartunian — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 4 files at r1, 12 of 12 files at r4, 2 of 8 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112865#-:-Ni0B7w_7mT8QloYCGre:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @kevin-v-ngo and @mgartner)

**Review by mgartner — APPROVED** [review]
LGTM

**Comment by maryliag** [comment]
TFTRs!
bors r+
